beacon_node/lighthouse_network/src/peer_manager/mod.rs

```diff
@@ -819,26 +833,290 @@ impl<TSpec: EthSpec> PeerManager<TSpec> {
+                    if let Some(peers_on_subnet) = subnet_to_peer.get_mut(&subnet_most_peers) {
+                        // and the subnet still contains peers
+                        if !peers_on_subnet.is_empty() {
+                            // Order the peers by the number of subnets they are long-lived
+                            // subscribed too, shuffle equal peers.
+                            peers_on_subnet.shuffle(&mut rand::thread_rng());
+                            peers_on_subnet.sort_by_key(|(_, info)| info.long_lived_subnet_count());
+
+                            // Try and find a candidate peer to remove from the subnet.
+                            // We ignore peers that would put us below our target outbound peers
+                            // and we currently ignore peers that would put us below our
+                            // sync-committee threshold, if we can avoid it.
+
+                            let mut removed_peer_index = None;
+                            for (index, (candidate_peer, info)) in
+                                peers_on_subnet.iter().enumerate()
+                            {
+                                // Ensure we don't remove too many outbound peers
+                                if info.is_outbound_only() {
+                                    if self.target_outbound_peers()
+                                        < connected_outbound_peer_count - outbound_peers_pruned
+                                    {
+                                        outbound_peers_pruned += 1;
+                                    } else {
+                                        // Restart the main loop with the outbound peer removed from
+                                        // the list. This will lower the peers per subnet count and
+                                        // potentially a new subnet may be chosen to remove peers. This
+                                        // can occur recursively until we have no peers left to choose
+                                        // from.
+                                        continue;
+                                    }
+                                }
+
+                                // Check the sync committee
+                                if let Some(subnets) = peer_to_sync_committee.get(candidate_peer) {
+                                    // The peer is subscribed to some long-lived sync-committees
+                                    // Of all the subnets this peer is subscribed too, the minimum
+                                    // peer count of all of them is min_subnet_count
+                                    if let Some(min_subnet_count) = subnets
+                                        .iter()
+                                        .map(|v| {
+                                            *sync_committee_peer_count.get(v).unwrap_or(&100u64)
+                                        })
```

Comment:
I think this could be cleaner as:

```rust
.filter_map(|v| sync_committee_peer_count.get(v).copied())
```

Alternatively we could defensively set the `unwrap_or` default to 0, so that the min is 0 and the peer isn't pruned if something has gone wrong with setting `sync_committee_peer_count`